### ISSUE TITLE
fooocus is very slow

### ISSUE THREAD

**Issue by JerryOver** [lllyasviel/Fooocus#473]
Fooocus is really great, but my A1111 is at least 10 times faster. When I look at the CMD window nothing happens for several minutes and then it'll spit the image out AT THAT POINT as fast as A1111. It's a great program, but it's very slow on my computer.

**Comment by barepixels** [comment]
Opposite here.  What kind of video card do you have?

**Comment by JerryOver** [comment]
RTX 3060. I watched a Youtube video since my post. It looked fast for the guy that I watched. I'm not sure why it's slow. My A1111 is pretty fast.

**Comment by barepixels** [comment]
3070 8 gig here

Using Fooocus-MRE version: 2.0.78.1 MRE  with these settings:

![a cat](https://github.com/lllyasviel/Fooocus/assets/92839033/3ad4c29f-6caa-45d1-928a-18918f6ec38c)

It took me  43.90 seconds

**Comment by lllyasviel** [comment]
newer comfyui is banned by colab so they sacrificed colab optimization to make 30xx series 3 to 4 seconds faster

but fooocus still have many colab users so we cannot 
follow that, official fooocus should be the current only software to use sdxl base and refiner on colab

if some forks of official fooocus has updated comfyui backend, then they are all broken on colab. some directly broken, some should broken after loading a refiner. 

if in the future, fooocus is also banned on colab, we will probably use similar method for speeding up things by sacrificing colab.

**Comment by lllyasviel** [comment]
regarding the issue of @JerryOver , please follow official installation guides and if issues are still here, paste the console outputs

**Comment by Reno0vacio** [comment]
> [JerryOver](/JerryOver)

Same here.. In  A1111 an image takes "max" 1 minute to render, depending on resolution etc..  but the same prompt, and the same resolution rendering on Fooocus is like, 3-4 minutes "at least". I updated my driver according to the README.dm to "513" version but still the same time. I have a 3060 card..

**Comment by MoonRide303** [comment]
@Reno0vacio Are you sure you have EXACTLY the same parameters in both A1111 and Fooocus, when comparing the speed? Same resolution, number of steps, sampler, scheduler? Using both base and refiner in A1111, or just base?

When not using refiner Fooocus is able to render image under 1 minute on 3050 (8 GB VRAM). First image using only base model took 1 minute, next image about 40 seconds. With refiner first image 95 seconds, next a bit under 60 seconds. Tested on [MRE](https://github.com/MoonRide303/Fooocus-MRE) v2.0.78.1, using relatively low-end PC with 16 GB of RAM. Total time values (pressing generate to seeing the image).

**Comment by barepixels** [comment]
Best test with a control test.  Prompt "A Cat"  everything else is set default in Fooocus and post result.  Oh fresh restart too

**Comment by JerryOver** [comment]
Yeah, I waited a week to see if there were any changes. I started over and downloaded the latest version of Foocus, Fooocus_win64_2-0-50. I even let it download the base models even though I had them. I even downloaded and installed the Nvidia Driver 531 because lllyasviel said that it may make it 10 times faster. It didn't. I prompted for 2 images of a car with refiner, and it took it 915 seconds. Without refiner, it took it 290 seconds. I will probably be using it now without refiner and using a model like Protovision that doesn't require it. The new output is down below. Thanks to all of those for their help so far.


D:\AI\Fooocus_win64_2-0-50>.\python_embeded\python.exe -s Fooocus\entry_with_update.py
Already up-to-date
Update succeeded.
Python 3.10.9 (tags/v3.10.9:1dd9be6, Dec  6 2022, 20:01:21) [MSC v.1934 64 bit (AMD64)]
Fooocus version: 2.0.78
Inference Engine exists.
Inference Engine checkout finished.
Total VRAM 12288 MB, total RAM 16099 MB
xformers version: 0.0.20
Set vram state to: NORMAL_VRAM
Device: cuda:0 NVIDIA GeForce RTX 3060 : native
Using xformers cross attention
Running on local URL:  http://127.0.0.1:7860

To create a public link, set `share=True` in `launch()`.
Fooocus Text Processing Pipelines are retargeted to cuda:0
[Virtual Memory System] Logic target is CPU, memory = 16098.671875
[Virtual Memory System] Activated = True
model_type EPS
adm 2560
making attention of type 'vanilla-xformers' with 512 in_channels
building MemoryEfficientAttnBlock with 512 in_channels...
Working with z of shape (1, 4, 32, 32) = 4096 dimensions.
making attention of type 'vanilla-xformers' with 512 in_channels
building MemoryEfficientAttnBlock with 512 in_channels...
BFloat16 VAE: Enabled
missing {'cond_stage_model.clip_g.transformer.text_model.embeddings.position_ids'}
Refiner model loaded: D:\AI\Fooocus_win64_2-0-50\Fooocus\models\checkpoints\sd_xl_refiner_1.0_0.9vae.safetensors
[Virtual Memory System] time = 1.06753s: model released from cpu: D:\AI\Fooocus_win64_2-0-50\Fooocus\models\checkpoints\sd_xl_refiner_1.0_0.9vae.safetensors
[Virtual Memory System] time = 0.28496s: refiner_clip released from cpu: D:\AI\Fooocus_win64_2-0-50\Fooocus\models\checkpoints\sd_xl_refiner_1.0_0.9vae.safetensors
model_type EPS
adm 2816
making attention of type 'vanilla-xformers' with 512 in_channels
building MemoryEfficientAttnBlock with 512 in_channels...
Working with z of shape (1, 4, 32, 32) = 4096 dimensions.
making attention of type 'vanilla-xformers' with 512 in_channels
building MemoryEfficientAttnBlock with 512 in_channels...
missing {'cond_stage_model.clip_g.transformer.text_model.embeddings.position_ids'}
Base model loaded: D:\AI\Fooocus_win64_2-0-50\Fooocus\models\checkpoints\sd_xl_base_1.0_0.9vae.safetensors
LoRAs loaded: [('sd_xl_offset_example-lora_1.0.safetensors', 0.5), ('None', 0.5), ('None', 0.5), ('None', 0.5), ('None', 0.5)]
Fooocus Expansion engine loaded for cuda:0.
App started successful. Use the app with http://127.0.0.1:7860/ or 127.0.0.1:7860
[Fooocus] Initializing ...
[Fooocus] Loading models ...
[Fooocus] Processing prompts ...
[Fooocus] Preparing Fooocus text #1 ...
[Prompt Expansion] New suffix: intricate, elegant, highly detailed, digital painting, artstation, concept art, smooth, sharp focus, illustration, art by artgerm and greg rutkowski and alphonse mucha and william - adolphe bouguereau
[Fooocus] Preparing Fooocus text #2 ...
[Prompt Expansion] New suffix: extremely detailed oil painting, unreal 5 render, rhads, Bruce Pennington, Studio Ghibli, tim hildebrandt, digital art, octane render, beautiful composition, trending on artstation, award-winning photograph, masterpiece
[Fooocus] Encoding base positive #1 ...
[Fooocus] Encoding base positive #2 ...
[Fooocus] Encoding base negative #1 ...
[Fooocus] Encoding base negative #2 ...
[Virtual Memory System] time = 1.28736s: refiner_clip loaded to cpu: D:\AI\Fooocus_win64_2-0-50\Fooocus\models\checkpoints\sd_xl_refiner_1.0_0.9vae.safetensors
[Fooocus] Encoding refiner positive #1 ...
[Fooocus] Encoding refiner positive #2 ...
[Fooocus] Encoding refiner negative #1 ...
[Fooocus] Encoding refiner negative #2 ...
[Virtual Memory System] time = 1.38372s: refiner_clip released from cpu: D:\AI\Fooocus_win64_2-0-50\Fooocus\models\checkpoints\sd_xl_refiner_1.0_0.9vae.safetensors
[ADM] Negative ADM = True
  0%|                                                                                           | 0/30 [00:00<?, ?it/s]D:\AI\Fooocus_win64_2-0-50\python_embeded\lib\site-packages\torchsde\_brownian\brownian_interval.py:594: UserWarning: Should have tb<=t1 but got tb=14.614643096923828 and t1=14.614643.
  warnings.warn(f"Should have {tb_name}<=t1 but got {tb_name}={tb} and t1={self._end}.")
 67%|██████████████████████████████████████████████████████▋                           | 20/30 [00:18<00:07,  1.34it/s][Virtual Memory System] time = 0.80218s: model released from cpu: D:\AI\Fooocus_win64_2-0-50\Fooocus\models\checkpoints\sd_xl_base_1.0_0.9vae.safetensors
[Virtual Memory System] time = 5.90171s: model loaded to cpu: D:\AI\Fooocus_win64_2-0-50\Fooocus\models\checkpoints\sd_xl_refiner_1.0_0.9vae.safetensors
Refiner swapped.
 93%|████████████████████████████████████████████████████████████████████████████▌     | 28/30 [01:06<00:03,  1.82s/it]D:\AI\Fooocus_win64_2-0-50\python_embeded\lib\site-packages\torchsde\_brownian\brownian_interval.py:585: UserWarning: Should have ta>=t0 but got ta=0.02916753850877285 and t0=0.029168.
  warnings.warn(f"Should have ta>=t0 but got ta={ta} and t0={self._start}.")
100%|██████████████████████████████████████████████████████████████████████████████████| 30/30 [01:07<00:00,  2.26s/it]
Diffusion time: 82.09 seconds
Image generated with private log at: D:\AI\Fooocus_win64_2-0-50\Fooocus\outputs\2023-10-02\log.html
[Virtual Memory System] time = 0.56097s: model released from cpu: D:\AI\Fooocus_win64_2-0-50\Fooocus\models\checkpoints\sd_xl_refiner_1.0_0.9vae.safetensors
[Virtual Memory System] time = 8.60502s: model loaded to cpu: D:\AI\Fooocus_win64_2-0-50\Fooocus\models\checkpoints\sd_xl_base_1.0_0.9vae.safetensors
 67%|██████████████████████████████████████████████████████▋                           | 20/30 [00:15<00:07,  1.34it/s][Virtual Memory System] time = 0.86003s: model released from cpu: D:\AI\Fooocus_win64_2-0-50\Fooocus\models\checkpoints\sd_xl_base_1.0_0.9vae.safetensors
[Virtual Memory System] time = 2.98698s: model loaded to cpu: D:\AI\Fooocus_win64_2-0-50\Fooocus\models\checkpoints\sd_xl_refiner_1.0_0.9vae.safetensors
Refiner swapped.
100%|██████████████████████████████████████████████████████████████████████████████████| 30/30 [01:02<00:00,  2.08s/it]
Diffusion time: 295.79 seconds
Image generated with private log at: D:\AI\Fooocus_win64_2-0-50\Fooocus\outputs\2023-10-02\log.html
Total time: 915.73 seconds
[Fooocus] Initializing ...
[Fooocus] Loading models ...
Refiner unloaded.
[Virtual Memory System] time = 7.04562s: model loaded to cpu: D:\AI\Fooocus_win64_2-0-50\Fooocus\models\checkpoints\sd_xl_base_1.0_0.9vae.safetensors
[Fooocus] Processing prompts ...
[Fooocus] Preparing Fooocus text #1 ...
[Prompt Expansion] New suffix: extremely high detail, solo, masterpiece, art by akihito tsukushi, alphonse mucha, kim jung gi, donato giancola, joseph christian leyendecker, raymond swanland, alphonse mucha, unreal engine 5
[Fooocus] Preparing Fooocus text #2 ...
[Prompt Expansion] New suffix: intricate, elegant, beautiful, highly detailed, incredible quality, trending on artstation
[Fooocus] Encoding base positive #1 ...
[Fooocus] Encoding base positive #2 ...
[Fooocus] Encoding base negative #1 ...
[Fooocus] Encoding base negative #2 ...
[ADM] Negative ADM = True
100%|██████████████████████████████████████████████████████████████████████████████████| 30/30 [00:22<00:00,  1.32it/s]
Diffusion time: 246.16 seconds
Image generated with private log at: D:\AI\Fooocus_win64_2-0-50\Fooocus\outputs\2023-10-02\log.html
100%|██████████████████████████████████████████████████████████████████████████████████| 30/30 [00:22<00:00,  1.35it/s]
Diffusion time: 26.75 seconds
Image generated with private log at: D:\AI\Fooocus_win64_2-0-50\Fooocus\outputs\2023-10-02\log.html
Total time: 290.11 seconds

**Comment by lllyasviel** [comment]
Given your log, your Fooocus spent 6 seconds to load model from your hard drive to system RAM, and then spent 30 seconds to generate each image on you 3060.

However, it wasted about 180 seconds to move the model from your system RAM to your 3060 GPU.

Your machine is special in that your GPU is good (3060) but your system RAM is old (16GB) and you are likely using old HDD instead of SSD hard drive (given the log).

The story is that, because your 16GB system RAM is not enough, Microsoft Windows Memory Swap started to use your hard disk as extra RAM. At the same time, your hard disk is relatively old HDD so the processing is very slow. (Maybe you have an SSD but that SSD does not have enough free space so your Windows must use that slow HDD as major processing unit.)

The solution is to load everything into your GPU and then completely avoid all model moving between GPU and system memory. To the best of my knowledge, Automatic1111 is the only software that is using this method. Completely avoid model moving usually makes things slower (because most users lack GPU VRAM rather than system RAM), but in some extremely special cases like this post, Automatic1111 is indeed faster.

In conclusion, for your machine the best solution is to use Automatic1111 because Comfy series rely heavily on system RAM.

**Comment by MoonRide303** [comment]
@JerryOver Like @lllyasviel pointed out it can be hard to make it any better on PC with slow drives, but you can try 2 more things:
1) Set Refiner to None before generation - it will lower quality a bit, but gives noticable performance boost in case of older hardware, especially machines with low RAM.
2) You can check out [Fooocus-MRE](https://github.com/MoonRide303/Fooocus-MRE) variant of mine - it uses newer ComfyUI backend that has few additional optimizations vs version used in vanilla (about 10% faster on my hardware, MRE v2.0.78.5 vs vanilla v2.0.78).

PS @lllyasviel ComfyUI got recently notable refactor of samplers and schedulers code - it should be easier to create custom sampler now (also via nodes), without having to patch it so much. It might be worth migrating to it and staying in sync - if not in V2, then in V3. It has to be done sooner or later, and IMO sooner is better - because for next generation of models (changes in unet code) we'll need to either update ComfyUI, or do even more of that ugly patching on Fooocus side (which should rather be cleaned up and replaced with proper pull requests to ComfyUI code base).

**Comment by JerryOver** [comment]
I'm glad @lllyasviel figured that out. I have two drives which are my C and D. I have been putting all my AI stuff on the D drive because it was bigger, so I checked to see what kind of drives that I have. My C is a SSD, but my D is a HDD. Tomorrow I'll put some time aside to set up my Fooocus again for my C Drive to see how it runs. Thanks so far.

**Comment by JerryOver** [comment]
Thanks a whole lot guys. That is what did it. I moved Fooocus to my C Drive which was SSD, and it took it 87 seconds to create a picture of a car with refiner. Thank you @lllyasviel.

**Comment by LIQUIDMIND111** [comment]
> I'm glad @lllyasviel figured that out. I have two drives which are my C and D. I have been putting all my AI stuff on the D drive because it was bigger, so I checked to see what kind of drives that I have. My C is a SSD, but my D is a HDD. Tomorrow I'll put some time aside to set up my Fooocus again for my C Drive to see how it runs. Thanks so far.

hello sir, how do you activate XFORMERS on FOOCUS? I only have them on auto1111 but i saw your LOG activates xformers on focus.... how?

**Comment by LIQUIDMIND111** [comment]
> @JerryOver Like @lllyasviel pointed out it can be hard to make it any better on PC with slow drives, but you can try 2 more things:
> 
> 1. Set Refiner to None before generation - it will lower quality a bit, but gives noticable performance boost in case of older hardware, especially machines with low RAM.
> 2. You can check out [Fooocus-MRE](https://github.com/MoonRide303/Fooocus-MRE) variant of mine - it uses newer ComfyUI backend that has few additional optimizations vs version used in vanilla (about 10% faster on my hardware, MRE v2.0.78.5 vs vanilla v2.0.78).
> 
> PS @lllyasviel ComfyUI got recently notable refactor of samplers and schedulers code - it should be easier to create custom samplers now (also via nodes), without having to patch it so much. It might be worth migrating to it and staying in sync - if not in V2, then maybe in V3? It has to be done sooner or later, and IMO sooner is better - because for next generation of models (changes in unet code) we'll need to either update ComfyUI, or do even more of that ugly patching on Fooocus side (which should rather be cleaned up and replaced with proper pull requests to ComfyUI code base).

what refiner do you recommend for people like me with 6VRAM and 16 RAM, i do get good speed even with 16 RAM, but o DONT use REFINER...... i love the output of juggernaut in REALISTIC BATCH MODE, but i dont use refiner... should i? woulkd it look even better than this that i made??? 
![image (24)](https://github.com/lllyasviel/Fooocus/assets/112510084/126f5d70-a664-4746-8fb2-760e38e38c38)

**Comment by welmentor** [comment]
> 3070 8 gigs card and 32 gigs RAM
> 
> Using Fooocus-MRE version: 2.0.78.1 with these settings:
> 
> ![a cat](https://private-user-images.githubusercontent.com/92839033/269831523-3ad4c29f-6caa-45d1-928a-18918f6ec38c.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDMzMzczNjMsIm5iZiI6MTcwMzMzNzA2MywicGF0aCI6Ii85MjgzOTAzMy8yNjk4MzE1MjMtM2FkNGMyOWYtNmNhYS00NWQxLTkyOGEtMTg5MThmNmVjMzhjLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFJV05KWUFYNENTVkVINTNBJTJGMjAyMzEyMjMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjMxMjIzVDEzMTEwM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTk4NzBjZjgxZjEzNTViNWNlNWNhZTc3NmNkZGZjMzc4MWZhNzMyMWViMTk5ZGEyYTM1OGI4YTU1ZTkxMWM0NTImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.M81B0vGw6rp7CCwpP-XU99I2XaB2Kig6G_g-yYcrygQ)
> 
> performance set to QUALITY Total time: 43.90 seconds
> 
> If I set performance to SPEED Total time: 26.99 seconds
> 
> =========================== Using Fooocus-lllyasviel version: 2.0.78 with ABOVE settings:
> 
> performance set to QUALITY Total time: 59.60 seconds
> 
> If I set performance to SPEED Total time: 31.59 seconds

----------
- Be thankful You guys are discussing in ''Seconds"
